src/test/java/org/jenkinsci/plugins/workflow/cps/steps/ParallelStepTest.java

```diff
@@ -517,4 +528,193 @@ public void parallelLexicalScope() throws Exception {
+        StringBuilder upstreamJobDefinition = new StringBuilder();
+        upstreamJobDefinition.append("node {\n");
+        upstreamJobDefinition.append("  parallel(\n");
+        for (FreeStyleProject downstreamJob : downstreamJobs) {
+            upstreamJobDefinition.append(
+                    String.format("    'branch-%s': {\n", downstreamJob.getName()));
+            upstreamJobDefinition.append(
+                    String.format("      semaphore 'sem-%s'\n", downstreamJob.getName()));
+            upstreamJobDefinition.append(
+                    String.format("      build '%s'\n", downstreamJob.getName()));
+            upstreamJobDefinition.append("    },\n");
+        }
+        upstreamJobDefinition.append("  )\n");
+        upstreamJobDefinition.append("}\n");
+        upstreamJob.setDefinition(new CpsFlowDefinition(upstreamJobDefinition.toString(), true));
+
+        WorkflowRun run = upstreamJob.scheduleBuild2(0).waitForStart();
+        for (int i = 0; i < downstreamJobs.size(); i++) {
+            FreeStyleProject downstreamJob = downstreamJobs.get(i);
+            Result downstreamResult = downstreamResults[i];
+
+            SemaphoreStep.success(String.format("sem-%s/1", downstreamJob.getName()), null);
+            story.j.waitForMessage(
+                    String.format("Scheduling project: %s", downstreamJob.getName()), run);
+            story.j.waitForMessage(
+                    String.format("Starting building: %s", downstreamJob.getName()), run);
+            List<FreeStyleBuild> downstreamBuilds = new ArrayList<>();
+            for (FreeStyleBuild downstreamBuild : downstreamJob.getBuilds()) {
+                downstreamBuilds.add(downstreamBuild);
+            }
+            assertEquals(1, downstreamBuilds.size());
+            story.j.waitForCompletion(downstreamBuilds.get(0));
+            if (!downstreamResult.equals(Result.SUCCESS)) {
+                story.j.waitForMessage(
+                        String.format("Failed in branch branch-%s", downstreamJob.getName()), run);
+            }
+        }
+
+        story.j.waitForCompletion(run);
+        story.j.assertBuildStatus(upstreamResult, run);
+
+        for (int i = 0; i < downstreamJobs.size(); i++) {
+            FreeStyleProject downstreamJob = downstreamJobs.get(i);
+            Result downstreamResult = downstreamResults[i];
+
+            if (!downstreamResult.equals(Result.SUCCESS)) {
+                story.j.waitForMessage(
+                        String.format(
+                                "%s #1 completed with status %s",
+                                downstreamJob.getName(), downstreamResult.toString()),
+                        run);
+            }
+        }
+    }
+
+    private FreeStyleProject getDownstreamJob(Result result) throws IOException {
+        FreeStyleProject ds = story.j.createFreeStyleProject();
+        ds.getBuildersList().add(new TestBuilder() {
```

Comment:
nit: Maybe simpler to use a Pipeline job for the downstream builds instead:

```
WorkflowJob ds = story.j.createProject(WorkflowJob.class);
ds.setDefinition(new CpsFlowDefinition("currentBuild.result = " + result, true));
```

Reply:
I don't think it makes a big difference one way or the other. I did it this way simply because I was copying an existing example from the `pipeline-build-step` plugin's tests.